peerinfo.go

```diff
@@ -31,25 +31,42 @@ func InfoFromP2pAddr(m ma.Multiaddr) (*PeerInfo, error) {
 		return nil, ErrInvalidAddr
 	}
 
-	ipfspart := parts[len(parts)-1] // last part
+	// TODO(lgierth): we shouldn't assume /ipfs is the last part
+	ipfspart := parts[len(parts)-1]
 	if ipfspart.Protocols()[0].Code != ma.P_IPFS {
 		return nil, ErrInvalidAddr
 	}
 
-	// make sure 'ipfs id' parses as a peer.ID
+	// make sure the /ipfs value parses as a peer.ID
 	peerIdParts := strings.Split(ipfspart.String(), "/")
 	peerIdStr := peerIdParts[len(peerIdParts)-1]
 	id, err := peer.IDB58Decode(peerIdStr)
 	if err != nil {
 		return nil, err
 	}
 
+	// we might have received just an /ipfs part, which means there's no addr.
+	var addrs []ma.Multiaddr
+	if len(parts) > 1 {
+		addrs = append(addrs, ma.Join(parts[:len(parts)-1]...))
+	}
+
 	return &PeerInfo{
 		ID:    id,
-		Addrs: []ma.Multiaddr{ma.Join(parts[:len(parts)-1]...)},
+		Addrs: addrs,
 	}, nil
 }
 
+func InfoToP2pAddrs(pi *PeerInfo) ([]ma.Multiaddr, error) {
+	addrs := []ma.Multiaddr{}
+	tpl := "/" + ma.ProtocolWithCode(ma.P_IPFS).Name + "/"
+	for _, addr := range pi.Addrs {
+		p2paddr := ma.StringCast(tpl + peer.IDB58Encode(pi.ID))
+		addrs = append(addrs, ma.Join(addr, p2paddr))
```

Comment:
It's not clear to me if there is a difference between `Join` and `Encapsulate` here, but if the result is the same I'd prefer `Encapsulate` (it's better doc'ed, and captures the meaning of the operation a bit better).